File: spring-reddit-clone/src/main/java/com/spring/redditclone/service/PostService.java
package com.spring.redditclone.service;

import com.spring.redditclone.model.PostRequest;
import com.spring.redditclone.model.PostResponse;

import java.util.List;

public interface PostService {
    PostResponse createPost(PostRequest request);
    List<PostResponse> getAllPosts();
    PostResponse getPost(Long id);
    List<PostResponse> getPostsBySubreddit(Long id);
    List<PostResponse> getPostsByUsername(String username);
}

File: frontend-reddit-clone-docker/src/app/shared/post.service.ts
import { PostRequest } from './../post/post-request.model';
import { Post } from './../home/post.model';
import { environment } from './../../environments/environment';
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class PostService {
  private apiUrl: string = environment.baseUrl;
  private postContext = '/posts';
  private postByUserContext = '/posts/by-user';

  constructor(private httpClient: HttpClient) { }

  public allPosts(): Observable<Post[]> {
    return this.httpClient.get<Post[]>(`${this.apiUrl}${this.postContext}`);
  }

  public createPost(postRequest: PostRequest): Observable<Post> {
    return this.httpClient.post<Post>(`${this.apiUrl}${this.postContext}`, postRequest);
  }

  public getPost(id: number): Observable<Post> {
    return this.httpClient.get<Post>(`${this.apiUrl}${this.postContext}/${id}`);
  }

  public getPostsByUser(username: string): Observable<Array<Post>> {
    return this.httpClient.get<Array<Post>>(`${this.apiUrl}${this.postByUserContext}/${username}`);
  }
}
